### DISCUSSION TITLE
Is there a way to access data points from previous sessions in an ongoing session?

### DISCUSSION THREAD

**Discussion by Nudelkoenig** [ctreffe/alfred#177] [Alfred Q&A] [answered]
I am conducting an experiment in which the participants estimate a number of values and I would like to give them feedback on how good their estimates are. Right now I show each individual particpant the average accuracy of their own estimates on a feedback page near the end of my experiment. But I would like to also show them how they stack up against previous participants. For this I would like to also show them the average accuracy of all previous participants. For this I would have to access the data from previous sessions while the feedback page is created.

So is there a way I can access the data of previous sessions during an ongoing session with **alfred3**?

**Comment by jobrachem** [comment, marked as answer]
Indeed, there is quite the convenient way. You can use the `ExperimentSession.all_exp_data` property ([docs](https://alfredo3.psych.bio.uni-goettingen.de/docs/generated/alfred3.experiment.ExperimentSession.all_exp_data.html)).

It returns a list of dictionaries. Each dictionary is the full dataset of an experiment session. Input elements are included with their names as keys and the user input as values. Here is a minimal example that demonstrates how to use it and how to explore it. *:warning: Note that the dictionary says `"mytextentry": None` here – even if you entered something on the previous page. Why? See below.*

```python
import pprint # for pretty printing of the list
import alfred3 as al
exp = al.Experiment()

exp += al.Page(name="p1")
exp.p1 += al.TextEntry(name="mytextentry")

@exp.member
class Demo(al.Page):
    
    def on_first_show(self):
        data = self.exp.all_exp_data
        pretty_data = pprint.pformat(data) # pretty printing

        self += al.CodeBlock(pretty_data, lang="python")
        

if __name__ == "__main__":
    exp.run()
```

The picture below shows the "Demo" page in the first session.

<img width="819" alt="image" src="https://user-images.githubusercontent.com/37882800/149371714-f52b968a-4a08-47e0-9a44-1e772ec1d73e.png">

### ⚠️  Important note
Data from the current session may take up to one second before appearing in this list. That means that if you use `ExperimentSession.all_exp_data` on Page 3 of an experiment, inputs made on Page 2 in the *current session* will most likely not be included in the corresponding dictionary, because they take up to 1 second before being written to the database. A safe way around this issue is: Use the `ExperimentSession.values` property for data from the current session instead. In this case, you should filter out the data of the current session from the list returned by the `ExperimentSession.all_exp_data` property, as shown below. *This is only relevant if there is any real danger that values you intend to use have been entered less than one seconds before you are accessing the data.*

```python
import pprint # for pretty printing of the list
import alfred3 as al
exp = al.Experiment()

exp += al.Page(name="p1")
exp.p1 += al.TextEntry(name="mytextentry")

@exp.member
class Demo(al.Page):
    
    def on_first_show(self):
        data = self.exp.all_exp_data
        data = [s for s in data if s["exp_session_id"] != self.exp.session_id] # filter data list
        pretty_data = pprint.pformat(data) # pretty printing

        self += al.CodeBlock(pretty_data, lang="python")
        

if __name__ == "__main__":
    exp.run()
```

**Comment by Nudelkoenig** [comment]
Thanks a lot! Works perfectly!
